dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/supplier/TrackedEntityProgramInstanceSupplier.java

```diff
@@ -0,0 +1,134 @@
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.tracker.preheat.supplier;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.program.ProgramInstance;
+import org.hisp.dhis.program.ProgramStatus;
+import org.hisp.dhis.tracker.TrackerImportParams;
+import org.hisp.dhis.tracker.domain.Enrollment;
+import org.hisp.dhis.tracker.preheat.TrackerPreheat;
+import org.hisp.dhis.tracker.util.Constant;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.stereotype.Component;
+
+import com.google.common.collect.Lists;
+
+/**
+ * Adds to the preheat a Map of Tracked Entities to related Program Instances
+ *
+ * @author Luca Cambi
+ */
+@Component
+public class TrackedEntityProgramInstanceSupplier extends JdbcAbstractPreheatSupplier
+{
+
+    private static final String PR_UID_COLUMN = "pr.uid";
+
+    private static final String PR_UID_COLUMN_ALIAS = "pruid";
+
+    private static final String PI_UID_COLUMN = "pi.uid";
+
+    private static final String PI_UID_COLUMN_ALIAS = "piuid";
+
+    private static final String PI_STATUS_COLUMN = "pi.status";
+
+    private static final String PI_STATUS_COLUMN_ALIAS = "status";
+
+    private static final String TEI_UID_COLUMN = "tei.uid";
+
+    private static final String TEI_UID_COLUMN_ALIAS = "teiuid";
+
+    private static final String SQL = "select  " + PR_UID_COLUMN + " as " + PR_UID_COLUMN_ALIAS + ", "
+        + PI_UID_COLUMN + " as " + PI_UID_COLUMN_ALIAS + ", "
+        + PI_STATUS_COLUMN + " as " + PI_STATUS_COLUMN_ALIAS + ", "
+        + TEI_UID_COLUMN + " as " + TEI_UID_COLUMN_ALIAS +
+        " from programinstance pi " +
+        " join trackedentityinstance tei on pi.trackedentityinstanceid = tei.trackedentityinstanceid " +
+        " join program pr on pr.programid = pi.programid " +
+        " where pi.deleted = false " +
+        " and tei.uid in (:uids)";
```

Comment:
Could we include a clause for programid as well?

"and pr.uid in (:programUids)"

Where programUids are the programs referenced in the payload (enrollments)? That way we reduce the number of rows returned even further :)

Reply:
ok @stian-sandvold i've updated it, we need to port it to 2.37 as it is already merged